plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/Inline.hs

```diff
@@ -407,3 +432,124 @@ maybeAddTySubst tn rhs = do
         modify' (extendType tn rhs)
         pure Nothing
     else pure $ Just rhs
+
+applyAndBetaReduce ::
+  forall tyname name uni fun ann.
+  (InliningConstraints tyname name uni fun) =>
+  -- | The rhs of the variable, should have been renamed already
+  Term tyname name uni fun ann ->
+  -- | The arguments, already processed
+  AppContext tyname name uni fun ann ->
+  InlineM tyname name uni fun ann (Term tyname name uni fun ann)
+applyAndBetaReduce rhs args = do
+  info <- ask
+  -- We run the beta and inlining passes for `len args` times, since sometimes the args
+  -- can only be beta-reduced one by one. An example is
+  -- `(\x y -> f x y) arg_x arg_y`, where `arg_x` and `arg_y` are effectful terms.
+  -- It needs two beta+inlining passes to reduce to `f arg_x arg_y`: the first pass
+  -- inlines `y` and the second pass `x`.
+  --
+  -- Note that the inlining passes we run here only run unconditional inlining.
+  -- Technically it can be helpful to also run callsite inlining, but it rarely
+  -- makes a difference and leads to exponential worst-case behavior.
```

Comment:
I'm not convinced that this isn't also exponential... why do we think that?

I notice that what we actually want to do is even simpler than running unconditional inlining: we want to do unconditional inlining, _only considering bindings from the current term_, and then just apply the substitution if we have one. I'm not sure if it's easy to express that, but it might make this less expensive.